pkg/lib/import-json.ts

```diff
@@ -0,0 +1,262 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2024 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <https://www.gnu.org/licenses/>.
+ */
+
+import { JsonValue, JsonObject } from "cockpit";
+
+/* GENERIC VALIDATION MACHINERY
+
+   This module helps with turning arbitrary user provided JSON blobs
+   into well-typed objects.
+
+   The basic idea is that for a TypeScript interface "Foo" you will
+   write a importer function with this signature:
+
+      function import_Foo(val: JsonValue): Foo;
+
+   This function will either return a valid Foo, or throw a
+   ValidationError.
+
+   When needing to convert a JSON blob into a Foo, you can call this
+   function directly. You might need to catch the potential
+   ValidationError.
+
+   Alternatively, you can also use the "validate" wrapper like so
+
+      const foo = validate("config.foo", config.foo, import_Foo, {});
+
+   This will include "config.foo" in the error messages to give a
+   better clue where the invalid data is actually coming from, and
+   will catch the ValidationError and return a fallback value.
+
+   Validation is generally lenient: If a validation error occurs deep
+   inside a nested structure, only the affected part of the structure
+   is omitted.  More conretely, if an element of an array is invalid,
+   this element is omitted from the array. If a optional field of an
+   object is invalid, it will be omitted. This doesn't happen
```

Comment:
This is what I mentioned above: IMHO this isn't safe in general. Think about `"bridges": []` in the shell manifest. Silently ignoring a non-string argument from sudo or pkexec is potentially disastrous. It's of course less of an issue for docs or locales, but IMHO it would make more sense to be strict and completely ignore the value -- that will greatly increase the chance of catching bugs in our tests.

(Also, "conretely" typo)

Reply:
Yeah, command lines are a very good example of something that better be rejected in its entirety.

Reply:
Put me also in the "reject outright" camp.

Reply:
Ok, convinced.

`import_Manifests` is a counter-example: We don't want to break all of Cockpit when one manifest is bad. But I guess we can easily write import_Manifests without import_record.